spec.emu

```diff
@@ -840,6 +843,7 @@ contributors: Nicolò Ribaudo
               1. <del>For each String _required_ of _module_.[[RequestedModules]], do</del>
               1. <ins>For each ModuleRequest Record _required_ of _module_.[[RequestedModules]], do</ins>
                 1. Let _requiredModule_ be GetImportedModule(_module_, _required_<ins>.[[Specifier]]</ins>).
+                1. <ins>If _required_.[[Phase]] is ~defer~ and _requiredModule_.GetExportedNames() is empty, throw a new *SyntaxError*.</ins>
```

Comment:
Note: other linking errors due to missing exports are also syntax errors.

Reply:
Typically named exports syntax errors apply to both the consumer and producer of a module, since on either side it is a syntax error not to define the correct export name.

In the case of defer, these are different though - the consumer has the intent of deferring execution, so it is not a syntax error if they don't realise the module has no exports. But from the perspective of the producer of the module being deferred it could be interpreted as a syntax error that they did not define export names to be deferred.

I do think perhaps the consumer use case wins here, in that a syntax error would be quite confusing to debug as a producer, while an explicit type error could make it clear that the "type of the module objet does not support defer".

Reply:
why would it not be a valid use case to defer import of a side-effecting zero-export module?

Reply:
It may be a valid use case, but the converse argument is yours in https://github.com/tc39/proposal-defer-import-eval/issues/19#issuecomment-1568779526.

Reply:
rofl indeed so.

is there a way to initiate the evaluation that's not accessing something on the namespace object? (since toStringTag and __esModule would be things we DONT want to trigger it)

Reply:
Not really, short of what Nicolo already landed in supporting `module.asdf` for any arbitrary string `asdf` as the executor. Unless we had something like `import.sync(module)`.

Reply:
i suppose `import(deferred)` would/could work already?

Reply:
That's not defined currently I believe. The goal for defer is to be used in all JS functions - sync and async, so dynamic import support isn't currently a feature.

Reply:
i'm not sure what you mean, `import()` works in any JS context?

i can't access the specifier from the deferred object, which might be passed around from its original import site, so that's not always an option.

Reply:
The defer use case is designed around synchronous lazy execution, not asynchronous lazy execution.

Reply:
🤔 

maybe `Object.keys(deferred)` and other forms of reflection could trigger it, via the MOP hooks?

Reply:
I'm not actually sure if this would work or not, perhaps the iteration is not execution triggering?

Reply:
Right now evaluation only happens during `[[Get]]`. It's possible to add it to more MOPs, but it makes it more difficult to teach why passing a ns object to some functions triggers evaluation and to other it doesn't.

`Object.keys` right now triggers evaluation only of non-empty namespaces, because it only calls `[[Get]]` if there are any own properties.

Reply:
Object.keys doesn't call [[Get]] at all, i think? Just Object.values/entries do - Object.keys calls [[OwnPropertyKeys]] and [[GetOwnProperty]]. I would expect all three of those to trigger evaluation.

Reply:
Yes, that would make sense to me, and I agree we shouldn't extend the behaviours unexpectedly here.

After some further thought I must also admit I'm coming around to the idea of permitting empty namespaces instead of throwing. Because it's an edge case, allowing it to work may be better for users than not.